src/llama-model.cpp

```diff
@@ -4345,6 +4372,103 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
                         layer.ffn_post_norm  = create_tensor(tn(LLM_TENSOR_FFN_POST_NORM, "weight", i), {n_embd}, 0);
                     }
                 } break;
+            case LLM_ARCH_GLM4_MOE:
+                {
+                    const int64_t n_expert        = hparams.n_expert;
+                    const int64_t n_expert_used   = hparams.n_expert_used;
+                    const int64_t n_expert_shared = hparams.n_expert_shared;
+
+                    tok_embd = create_tensor(tn(LLM_TENSOR_TOKEN_EMBD, "weight"), { n_embd, n_vocab }, 0);
+
+                    // output
+                    output_norm = create_tensor(tn(LLM_TENSOR_OUTPUT_NORM, "weight"), { n_embd }, 0);
+                    output = create_tensor(tn(LLM_TENSOR_OUTPUT, "weight"), { n_embd, n_vocab }, TENSOR_NOT_REQUIRED);
+                    // if output is NULL, init from the input tok embed
+                    if (output == NULL) {
+                        output = create_tensor(tn(LLM_TENSOR_TOKEN_EMBD, "weight"), { n_embd, n_vocab }, TENSOR_DUPLICATED);
+                    }
+
+                    // NextN/MTP tensors (preserved but unused) - only in final layer (46 for Air, 92 for GLM-4.5)
+                    const int final_layer = n_layer - 1; // NextN tensors are in the last layer only
+                    create_tensor(tn(LLM_TENSOR_NEXTN_EH_PROJ, final_layer), { 2 * n_embd, n_embd }, TENSOR_NOT_REQUIRED);
+                    create_tensor(tn(LLM_TENSOR_NEXTN_EMBED_TOKENS, final_layer), { n_embd, n_vocab }, TENSOR_NOT_REQUIRED);
+                    create_tensor(tn(LLM_TENSOR_NEXTN_ENORM, final_layer), { n_embd }, TENSOR_NOT_REQUIRED);
+                    create_tensor(tn(LLM_TENSOR_NEXTN_HNORM, final_layer), { n_embd }, TENSOR_NOT_REQUIRED);
+                    create_tensor(tn(LLM_TENSOR_NEXTN_SHARED_HEAD_HEAD, final_layer), { n_embd, n_vocab }, TENSOR_NOT_REQUIRED);
+                    create_tensor(tn(LLM_TENSOR_NEXTN_SHARED_HEAD_NORM, final_layer), { n_embd }, TENSOR_NOT_REQUIRED);
+
+                    for (int i = 0; i < n_layer; ++i) {
```

Comment:
A bit of a guess here but should the last layer be used in this loop when not doing MTP? And just loop to n_layer-1. Since it gets treated the same as all other layers (same goes for the other similar loop in the code). 

Looking at the vllm implementation, they have separated the handling of these layers by checking if it is a special layer, `spec_layer = get_spec_layer_idx_from_weight_name(self.config, name)` and treat it differently. 

And as someone mentioned earlier to look at the Transformers implementation, they are also skipping that last layer when initiating the model  
`self.layers = nn.ModuleList([Glm4MoeDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)])` where config.num_hidden_layers is 46 while our n_layer is set to 47.

Reply:
When I looked at layer 46, it contained:

1. COMPLETE Regular Transformer Tensors:

- Attention: attn_q, attn_k, attn_v, attn_output, all biases, attn_norm
- MoE FFN: ffn_gate_inp, ffn_*_exps, ffn_*_shexp, ffn_norm, exp_probs_b

2. NextN/MTP Tensors (unused in inference):

- eh_proj, embed_tokens, enorm, hnorm, shared_head.*

So my conclusion here was:

Layer 46 is a complete transformer layer that should be processed in the forward pass. It has:
- Full attention mechanism
- Full MoE FFN (128 experts + shared expert)
- Expert routing tensors

Therefor the NextN tensors are additional and are marked as TENSOR_NOT_REQUIRED (unused).

Or have I missed something?

Reply:
I think the MTP layer issues are now resolved, please re-open if that's not the case.